diff --git a/vllm/model_executor/layers/fused_moe/layer.py b/vllm/model_executor/layers/fused_moe/layer.py
index 12301ade2..c188c77b5 100644
--- a/vllm/model_executor/layers/fused_moe/layer.py
+++ b/vllm/model_executor/layers/fused_moe/layer.py
@@ -1549,8 +1549,7 @@ class FusedMoE(CustomOp):
 
         quant_method_name = self.quant_method.__class__.__name__
         global_expert_id = expert_id
-        expert_id = self._map_global_expert_id_to_local_expert_id(
-            global_expert_id)
+        expert_id = self._map_global_expert_id_to_local_expert_id(global_expert_id)
         is_modeloptnvfp4 = quant_method_name == "ModelOptNvFp4FusedMoE"
         is_input_scale = "input_scale" in weight_name
         if expert_id == -1 and not (is_modeloptnvfp4 and is_input_scale):
@@ -1644,7 +1643,8 @@ class FusedMoE(CustomOp):
             self._load_single_value(
                 param=param,
                 loaded_weight=loaded_weight,
-                expert_id=global_expert_id if is_modeloptnvfp4 else expert_id)
+                expert_id=global_expert_id if is_modeloptnvfp4 else expert_id,
+            )
             return True if return_success else None
 
         # Case g_idx
diff --git a/vllm/model_executor/layers/quantization/modelopt.py b/vllm/model_executor/layers/quantization/modelopt.py
index fb968ebf8..fdd3ad439 100644
--- a/vllm/model_executor/layers/quantization/modelopt.py
+++ b/vllm/model_executor/layers/quantization/modelopt.py
@@ -1301,14 +1301,16 @@ class ModelOptNvFp4FusedMoE(FusedMoEMethodBase):
             {"quant_method": FusedMoeWeightScaleSupported.TENSOR.value}
         )
 
-        w13_input_scale = PerTensorScaleParameter(data=torch.empty(
-            global_num_experts, 2, dtype=torch.float32),
-                                                  weight_loader=weight_loader)
+        w13_input_scale = PerTensorScaleParameter(
+            data=torch.empty(global_num_experts, 2, dtype=torch.float32),
+            weight_loader=weight_loader
+        )
         layer.register_parameter("w13_input_scale", w13_input_scale)
 
-        w2_input_scale = PerTensorScaleParameter(data=torch.empty(
-            global_num_experts, dtype=torch.float32),
-                                                 weight_loader=weight_loader)
+        w2_input_scale = PerTensorScaleParameter(
+            data=torch.empty(global_num_experts, dtype=torch.float32),
+            weight_loader=weight_loader
+        )
 
         layer.register_parameter("w2_input_scale", w2_input_scale)
 
@@ -1462,8 +1464,9 @@ class ModelOptNvFp4FusedMoE(FusedMoEMethodBase):
         layer.w13_weight_scale_2 = Parameter(w13_weight_scale_2, requires_grad=False)
 
         # Common processing for input scales and alphas
-        w13_input_scale = layer.w13_input_scale.max().to(torch.float32).expand(
-            layer.num_experts)
+        w13_input_scale = (
+            layer.w13_input_scale.max().to(torch.float32).expand(layer.num_experts)
+        )
         layer.g1_alphas = Parameter(
             (w13_input_scale * w13_weight_scale_2).to(torch.float32),
             requires_grad=False,
@@ -1475,15 +1478,18 @@ class ModelOptNvFp4FusedMoE(FusedMoEMethodBase):
         )
 
         # GEMM 2 processing
-        w2_input_scale = layer.w2_input_scale.max().to(torch.float32).expand(
-            layer.num_experts)
+        w2_input_scale = (
+            layer.w2_input_scale.max().to(torch.float32).expand(layer.num_experts)
+        )
         layer.g2_alphas = Parameter(
             (w2_input_scale * layer.w2_weight_scale_2).to(torch.float32),
-            requires_grad=False)
+            requires_grad=False,
+        )
 
         # This is for quantization, so we need to invert it.
         layer.w2_input_scale_quant = Parameter(
-            (1 / w2_input_scale).to(torch.float32), requires_grad=False)
+            (1 / w2_input_scale).to(torch.float32), requires_grad=False
+        )
 
         # TensorRT-LLM specific processing
         if (
